google/cloud/security/common/gcp_type/firewall_rule.py

```diff
@@ -65,11 +83,189 @@ def __init__(self, **kwargs):
+
+        Required fields: name and network
+        Length restrictions:
+          * name <= 63 characters
+          * <= 256 values:
+            sourceRanges, sourceTags, targetTags, destinationRanges
+
+        Raises:
+          InvalidFirewallRuleError: If keys don't meet requirements.
+        """
+        if not self.name:
+            raise InvalidFirewallRuleError(
+                'Rule missing required field "%s"' % 'name')
+        if not self.network:
+            raise InvalidFirewallRuleError(
+                'Rule missing required field "%s"' % 'network')
+
+        if len(self.name) > 63:
+            raise InvalidFirewallRuleError(
+                'Rule name exceeds length limit of 63 chars: "%s".' %
+                self.name)
+
+        max_256_value_keys = [
+            ('sourceRanges', self._source_ranges),
+            ('sourceTags', self._source_tags),
+            ('targetTags', self._target_tags),
+            ('destinationRanges', self._destination_ranges)
+        ]
+        for key, value in max_256_value_keys:
+            if value and len(value) > 256:
+                raise InvalidFirewallRuleError(
+                    'Rule entry "%s" must contain 256 or fewer values: "%s".'
+                    % (key, value))
+
+        if self._source_tags:
+            if self._source_service_accounts or self._target_service_accounts:
+                raise InvalidFirewallRuleError(
+                    'sourceTags cannot be set when source/targetServiceAccounts'
+                    ' are set')
+
+        if self._target_tags:
+            if self._source_service_accounts or self._target_service_accounts:
+                raise InvalidFirewallRuleError(
+                    'sourceTags cannot be set when source/targetServiceAccounts'
+                    ' are set')
+
+        max_1_value_keys = [
+            ('sourceServiceAccount', self.source_service_accounts),
+            ('targetServiceAccount', self.target_service_accounts),
+        ]
+        for key, value in max_1_value_keys:
+            if value and len(value) > 1:
+                raise InvalidFirewallRuleError(
+                    'Rule entry "%s" may contain at most 1 value: "%s".'
+                    % (key, value))
+
+    def _validate_direction(self):
+        """Checks that the direction and associated fields are valid.
+
+        Raises:
+          InvalidFirewallRuleError: If:
+            * Direction is 'ingress' and
+              * there are no source ranges or tags
+              * _destination_ranges is not set
+            * Direction is 'egress' and
+              * there are no source ranges or tags
+              * _destination_ranges is set
+        """
+        if self.direction == 'ingress':
+            if (not self._source_ranges and not self._source_tags and not
+                    self.source_service_accounts):
+                raise InvalidFirewallRuleError(
+                    'Ingress rule missing required field oneof '
```

Comment:
Update message to include sourceServiceAccounts

Reply:
Done.